rfd/9999-aws-api-integration-using-oidc.md

```diff
@@ -0,0 +1,621 @@
+	"subkindData": {
+		"aws_role_arn": "arn:aws:123:TeleportOIDC"
+	}
+}
+```
+
+In order to call an Integration action, we'll create a new Web API endpoint:
+```
+--->
+POST .../integration/:id/action/aws-oidc-list-databases
+{}
+
+<---
+{
+    "status": "success",
+    "response": {
+        "items": [
+            {
+                "status": "available",
+                "name": "marcodbtest",
+                "iamAuth": "true",
+                "engine": "postgres",
+                "engineVersion": "15.2",
+                "masterUsername": "postgres",
+                "tags": [],
+                "arn": "arn:aws:rds:us-east-1:<accid>:db:marcodbtest",
+                "addr": "marcodbtest.<someid>.us-east-1.rds.amazonaws.com",
+                "port": "5432"
+            }
+        ]
+    }
+}
+```
+
+This will generate a new token for the Integration `:id` and then call the `aws-oidc-list-databases` method which maps to a method that will do two AWS API Calls:
+- `rds describe-db-instances`
+- `rds describe-db-clusters`
+
+The response contains the following fields:
+- `status`: reports the result of calling the integration (other errors will come as usual: http status code)
+- `response`: specific to the requested action
+
+Pagination/Offset and search capabilities are out of scope for now.
+
+#### CLI
+Users can also create/update the resource using `tctl`.
+
+The only updatable field is `aws_role_arn`.
+
+`tctl`:
+
+```
+$ tctl get integrations
+kind: integration
+subkind: aws-oidc
+version: v1
+metadata:
+	name: some-name
+spec:
+	aws_role_arn: arn:aws:123:TeleportOIDC
+
+$ tctl get integration/myaws
+kind: integration
+subkind: aws-oidc
+version: v1
+metadata:
+	name: some-name
+spec:
+	aws_role_arn: arn:aws:123:TeleportOIDC
+
+$ tctl create aws-integration.yaml
+```
+
+Resource representation as `yaml`:
+```yaml
+kind: integration
+subkind: aws-oidc
+version: v1
+metadata:
+	name: myaws
+spec:
+	aws_role_arn: arn:aws:123:TeleportOIDC
+```
+
+#### IaC - Terraform
+A new resource `Integration` must be created to allow for the Integration management from the Terraform provider.
+
+#### IaC - Kube Operator
+A new resource `Integration` must be created to allow for the Integration management from the Kube Operator.
+
+#### IaC - Helm Charts
+No change will be made to the Helm Charts because there's no new configuration changes.
+
+### Security
+
+#### Rotation
+JWT Signing keys can be rotated using `tctl auth rotate --type=oidc`.
+
+JWKS endpoint returns a list of public keys, so the old and the new key are provided for a seamless migration.
+
+#### MITM between AWS and Teleport
+
+The only configuration AWS has about the provider is its URL.
+If DNS is controlled by an attacker, AWS might accept requests from them.
+
+However, assuming the DNS is controlled by an attacker, then the users that are logging in to Teleport are also providing their credentials to that part.
+
+We don't think this is a scenario we should focus on this RFD.
```

Comment:
This is somewhat mitigated by AWS's thumb printing of the ID provider's certificate's top intermediate certificate authority - an attacker would need to poison AWS's DNS resolvers and also obtain a legitimate certificate issued by the same intermediate CA.

I agree with your overall assessment here. If AWS's DNS resolvers have been poisoned and their chain of trust compromised, that's a much bigger problem that's out of our hands no matter what we do.

**We should really add a warning that if you change your certificate provider that issues your Teleport proxy certificate, a new fingerprint must be added to AWS.**

Reply:
Definitely. I noticed this earlier, from the AWS IdP Page

> Server certificate thumbprint is the hex-encoded SHA-1 hash value of the X.509 certificate used by the domain where the OpenID Connect provider makes its keys available.

We should aim to mitigate this somehow, be it via warning or an alert

Reply:
I added some notes here
https://github.com/gravitational/teleport/pull/22556/commits/583d8b3f8d18ad7678f6c1b7848505c23cc402c9

> be it via warning or an alert

Teleport alerting the user?
:thought_balloon: 
We can store our own thumbprint within the integration.
If we get an error from AWS (403?), we validate if the thumbprint is still the same.
If they are not the same, then warn the user about it (and that they need to set up the integration again).

Reply:
I think a warning in the docs somewhere may suffice for now, but, if we were able to prompt in the logs on failure that would be awesome :')

Reply:
I was thinking maybe a cluster alert if the fingerprint has changed